rpc/src/server.rs

```diff
@@ -23,93 +34,148 @@ impl RpcServer {
-                            context.sender.clone(),
-                        )))
-                    },
-                )
-                .start(
-                    &tcp_listen_address
-                        .to_socket_addrs()
-                        .expect("config tcp_listen_address parsed")
-                        .next()
-                        .expect("config tcp_listen_address parsed"),
-                )
-                .expect("Start Jsonrpc TCP service");
-                info!("Listen TCP RPCServer on address {}", tcp_listen_address);
-
-                tcp_server
+    /// * `handler` - Tokio runtime handle.
+    pub fn new(config: RpcConfig, io_handler: IoHandler, handler: Handle) -> Self {
+        let rpc = Arc::new(io_handler);
+
+        let http_address = Self::start_server(
+            &rpc,
+            config.listen_address.to_owned(),
+            handler.clone(),
+            false,
+        )
+        .map(|local_addr| {
+            info!("Listen HTTP RPCServer on address: {}", local_addr);
+            local_addr
+        })
+        .unwrap();
+
+        let ws_address = if let Some(addr) = config.ws_listen_address {
+            let local_addr = Self::start_server(&rpc, addr, handler.clone(), true).map(|addr| {
+                info!("Listen WebSocket RPCServer on address: {}", addr);
+                addr
             });
+            local_addr.ok()
+        } else {
+            None
+        };
 
-        let _ws = config.ws_listen_address.as_ref().map(|ws_listen_address| {
-            let subscription_rpc_impl = SubscriptionRpcImpl::new(notify_controller.clone(), handle);
-            let mut handler = io_handler.clone();
-            if config.subscription_enable() {
-                handler.extend_with(subscription_rpc_impl.to_delegate());
-            }
-            let ws_server = jsonrpc_ws_server::ServerBuilder::with_meta_extractor(
-                handler,
-                |context: &jsonrpc_ws_server::RequestContext| {
-                    Some(SubscriptionSession::new(Session::new(context.sender())))
-                },
-            )
-            .start(
-                &ws_listen_address
+        let tcp_address = if let Some(addr) = config.tcp_listen_address {
+            let local_addr = handler.block_on(Self::start_tcp_server(rpc, addr));
+            if let Ok(addr) = &local_addr {
+                info!("Listen TCP RPCServer on address: {}", addr);
+            };
+            local_addr.ok()
+        } else {
+            None
+        };
+
+        Self {
+            http_address,
+            tcp_address,
+            ws_address,
+        }
+    }
+
+    fn start_server(
+        rpc: &Arc<MetaIoHandler<Option<Session>>>,
+        address: String,
+        handler: Handle,
+        enable_websocket: bool,
+    ) -> Result<SocketAddr, AnyError> {
+        let stream_config = StreamServerConfig::default()
+            .with_channel_size(4)
+            .with_pipeline_size(4);
```

Comment:
Do these configurations have an impact on performance? Should we consider making these parameters customizable?

Reply:
these parameter only have effect on websockets.